numpy/_core/src/multiarray/descriptor.c

```diff
@@ -1813,15 +1813,30 @@ _convert_from_str(PyObject *obj, int align)
         int kind;
 
         /* Parse the integer, make sure it's the rest of the string */
-        elsize = (int)strtol(type + 1, &typeend, 10);
-        /* Make sure size is not negative */
-        if (elsize < 0) {
+        long result = strtol(type + 1, &typeend, 10);
+        // make sure it doesn't overflow or go negative
+        if (result > INT_MAX || result < 0) {
             goto fail;
         }
 
+        elsize = (int)result;
+
+
         if (typeend - type == len) {
 
             kind = type[0];
+            // if it's a type + size dtype and the parsed size is zero
+            // and the size of the typestring is longer than 1 character
+            // then strtol must have returned 0 indicating an error
+            if ((elsize == 0) && (typeend > type + 2) &&
```

Comment:
Would `S00000` now fail? From the description of `strtol`, it also seems that in your example of `S123hello`, `typeend - type` would not be `len`.

Reply:
Looks like it:

```
In [1]: np.dtype('S0000000')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 np.dtype('S0000000')

TypeError: data type 'S0000000' not understood
```

Let me see if I can find a better heuristic that works with that test case...

Reply:
Ended up deciding the `typeend - type == len` check is sufficient and getting rid of the `strtol` error checking.